nni/nas/space/space.py

```diff
@@ -0,0 +1,331 @@
+    """
+
+    status: ModelStatus
+    """The status of the model space / model."""
+
+    metrics: Metrics
+    """The evaluation metrics of the model."""
+
+    evaluator: Evaluator | None
+    """Evaluator that assesses the quality of the model."""
+
+    sample: Sample | None
+    """The sample that is used to freeze this model. It's useful for debug and visualization.
+    It could be left unset if sample is not used when freezing the model.
+    """
+
+    def __init__(self, status: ModelStatus = ModelStatus.Initialized) -> None:
+        self.status = status
+        self.metrics = Metrics()
+
+    def execute(self) -> Any:
+        """Execute the training (and/or evaluation)."""
+        if self.evaluator is None:
+            raise ValueError('Evaluator is not set, but default execute requires an evaluator.')
+        executable_model = self.executable_model()
+        return self.evaluator.evaluate(executable_model)
+
+    @classmethod
+    def from_model(cls, model_space: BaseModelSpace, evaluator: Evaluator | None = None, **configs: Any) -> ExecutableModelSpace:
+        """Convert any model space to a specific type of executable model space.
+
+        Parameters
+        ----------
+        model_space
+            Model space written in deep learning framework in most cases.
+        evaluator
+            A model usually requires an evaluator to be *executable*.
+            But evaluator can sometimes be optional for debug purposes or to support fancy algorithms.
+        configs
+            Additional configurations for the executable model space.
+
+        Returns
+        -------
+        The converted model space.
+        """
+        raise NotImplementedError('`from_model` is not implemented for {}'.format(cls.__name__))
+
+    def executable_model(self) -> Any:
+        """Fully instantiate the deep learning model (e.g., PyTorch Module) so that it's ready to be executed.
+
+        :meth:`executable_model` is usually symmetrical to :meth:`from_model`.
+        While :meth:`from_model` converts deep learning model to :class:`ExecutableModelSpace`,
+        :meth:`executable_model` converts :class:`ExecutableModelSpace` back to deep learning model.
+
+        Returns
+        -------
+        Typical this method should return a PyTorch / Tensorflow model (or model factory),
+        depending on the input format of evaluator.
+        """
+        raise NotImplementedError('`executable_model` is not implemented for {}'.format(self.__class__.__name__))
+
+    @property
+    def metric(self) -> float | None:
+        """Training result of the model, or ``None`` if it's not yet trained or has failed to train."""
+        result = self.metrics.final
+        # Didn't check the status here. Even if the model failed to train, it can still have metrics sometimes.
+        if isinstance(result, dict):
+            return result.get('default', None)
+        return result
+
+
+class KeepModelSpace(ExecutableModelSpace):
```

Comment:
not clear what is this class used for...

Reply:
I can't remind. Did we determined the class name last time?

Reply:
I came up with a new name:

RawFormatModelSpace